File: frontend/src/lib/jobExtraction.ts
import type { Job, JobExtraction } from "../types";

const MIN_EXTRACT_CHARS = 100;

export function extractionMetadata(
  extraction: JobExtraction,
  jobText?: string,
): Record<string, unknown> {
  return {
    employment_type: extraction.employment_type ?? null,
    salary_range: extraction.salary_range ?? null,
    requirements: extraction.requirements,
    work_mode: extraction.work_mode ?? null,
    match_summary: extraction.match_summary,
    ...(jobText ? { job_text: jobText } : {}),
  };
}

export function buildJobExtractSource(job: Pick<Job, "title" | "company" | "location" | "description" | "raw_metadata">): string {
  const stored = job.raw_metadata?.job_text;
  if (typeof stored === "string" && stored.trim().length >= MIN_EXTRACT_CHARS) {
    return stored.trim();
  }

  const header = [job.title, [job.company, job.location].filter(Boolean).join(" · ")].filter(Boolean).join("\n");
  return [header, "", job.description].join("\n").trim();
}

export function canExtractFromText(text: string): boolean {
  return text.trim().length >= MIN_EXTRACT_CHARS;
}

File: app/schemas/job_extraction.py
from typing import Literal

from pydantic import BaseModel, Field

WorkMode = Literal["remote", "hybrid", "on-site", "flexible"]


class JobExtraction(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    company: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1)
    work_mode: WorkMode | None = None
    location: str | None = Field(default=None, max_length=255)
    employment_type: str | None = Field(default=None, max_length=100)
    salary_range: str | None = Field(default=None, max_length=255)
    requirements: list[str] = Field(default_factory=list)
